src/pks/PK_Helpers.cc

```
@@ -0,0 +1,288 @@
+{
+  if (u.HasComponent("face")) {
+    Epetra_MultiVector& u_f = *u.ViewComponent("face", false);
+    for (unsigned int f = 0; f != u_f.MyLength(); ++f) {
+      if (bcs.bc_model()[f] == Operators::OPERATOR_BC_DIRICHLET) { u_f[0][f] = bcs.bc_value()[f]; }
+    }
+
+  } else if (u.HasComponent("boundary_face")) {
+    Epetra_MultiVector& u_bf = *u.ViewComponent("boundary_face", false);
+    const Epetra_MultiVector& u_c = *u.ViewComponent("cell", false);
+    const Epetra_Map& vandalay_map =
+      u.Mesh()->getMap(AmanziMesh::Entity_kind::BOUNDARY_FACE, false);
+    const Epetra_Map& face_map = u.Mesh()->getMap(AmanziMesh::Entity_kind::FACE, false);
+
+    for (int bf = 0; bf != u_bf.MyLength(); ++bf) {
+      AmanziMesh::Entity_ID f = face_map.LID(vandalay_map.GID(bf));
+      if (bcs.bc_model()[f] == Operators::OPERATOR_BC_DIRICHLET) {
+        u_bf[0][bf] = bcs.bc_value()[f];
+      }
+    }
+  }
+}
+
+
+// -----------------------------------------------------------------------------
+// Given a vector and a face ID, get the value at that location.
+//
+// Looks in the following order:
+//  -- face component
+//  -- boundary Dirichlet data
+//  -- boundary_face value
+//  -- internal cell
+// -----------------------------------------------------------------------------
+double
+getFaceOnBoundaryValue(AmanziMesh::Entity_ID f, const CompositeVector& u, const Operators::BCs& bcs)
+{
+  if (u.HasComponent("face")) {
+    return (*u.ViewComponent("face", false))[0][f];
+  } else if (bcs.bc_model()[f] == Operators::OPERATOR_BC_DIRICHLET) {
+    return bcs.bc_value()[f];
+    // } else if (u.HasComponent("boundary_face")) {
+    //   AmanziMesh::Entity_ID bf = getFaceOnBoundaryBoundaryFace(*u.Mesh(), f);
+    //   return (*u.ViewComponent("boundary_face",false))[0][bf];
+  } else {
+    auto c = getFaceOnBoundaryInternalCell(*u.Mesh(), f);
+    return (*u.ViewComponent("cell", false))[0][c];
+  }
+  return -1;
```

If we are here, we should abort instead of returning a meaningful value.